ecs-cli/modules/cli/servicediscovery/servicediscovery_helper.go

```diff
@@ -0,0 +1,290 @@
+		return fmt.Errorf("Validation Error: %s and %s both specified", flags.PrivateDNSNamespaceIDFlag, flags.PrivateDNSNamespaceNameFlag)
+	}
+
+	if c.String(flags.PublicDNSNamespaceIDFlag) != "" && c.String(flags.PublicDNSNamespaceNameFlag) != "" {
+		return fmt.Errorf("Validation Error: %s and %s both specified", flags.PublicDNSNamespaceIDFlag, flags.PublicDNSNamespaceNameFlag)
+	}
+
+	if ecsParamsSD.PrivateDNSNamespace.Name != "" && ecsParamsSD.PrivateDNSNamespace.ID != "" {
+		return fmt.Errorf("Validation Error: private_dns_namespace.name and private_dns_namespace.id both specified")
+	}
+
+	if ecsParamsSD.PublicDNSNamespace.Name != "" && ecsParamsSD.PublicDNSNamespace.ID != "" {
+		return fmt.Errorf("Validation Error: public_dns_namespace.name and public_dns_namespace.id both specified")
+	}
+
+	return nil
+}
+
+// Logs warnings for fields which are ignored when a namespace ID is specified
+func namespaceWarningsWhenIDSpecified(input *utils.ServiceDiscovery) {
+	msg := "Ignoring %s because the ID of an existing private namespace was specified"
+
+	privNamespace := input.PrivateDNSNamespace
+	if privNamespace.ID != "" {
+		if privNamespace.Description != "" {
+			logrus.Warnf(msg, "description")
+		}
+		if privNamespace.VPC != "" {
+			logrus.Warnf(msg, "vpc")
+		}
+	}
+}
+
+func hasNamespace(n utils.Namespace) bool {
+	if n.ID != "" || n.Name != "" {
+		return true
+	}
+	return false
+}
+
+func getNamespace(n utils.Namespace) string {
+	if n.ID != "" {
+		return n.ID
+	}
+	return n.Name
+}
+
+func usesSRV(input *utils.ServiceDiscovery, networkMode string) bool {
+	dnsType := input.ServiceDiscoveryService.DNSConfig.Type
+	if dnsType == servicediscovery.RecordTypeSrv {
+		return true
+	}
+	if dnsType == "" && (networkMode == ecs.NetworkModeHost || networkMode == ecs.NetworkModeBridge) {
+		return true
+	}
+	return false
+}
+
+func getOutputIDFromStack(cfnClient cloudformation.CloudformationClient, stackName, outputKey string) (*string, error) {
+	response, err := cfnClient.DescribeStacks(stackName)
+	if err != nil {
+		return nil, err
+	}
+	if len(response.Stacks) == 0 {
+		return nil, fmt.Errorf("Could not find CloudFormation stack: %s", stackName)
+	}
+
+	for _, output := range response.Stacks[0].Outputs {
+		if aws.StringValue(output.OutputKey) == outputKey {
+			return output.OutputValue, nil
+		}
+	}
+	return nil, fmt.Errorf("Failed to find output %s in stack %s", outputKey, stackName)
+}
+
+func cfnStackName(stackName, cluster, service string) string {
+	return fmt.Sprintf(stackName, cluster, service)
+}
+
+func getSDSCFNParams(namespaceID, sdsName, networkMode string, input *utils.ServiceDiscovery) *cloudformation.CfnStackParams {
+	cfnParams := cloudformation.NewCfnStackParams(requiredParamsSDS)
+
+	cfnParams.Add(parameterKeyNamespaceID, namespaceID)
+	cfnParams.Add(parameterKeySDSName, sdsName)
+
+	dnsType := getDNSType(input, networkMode)
```

Comment:
Just noting that `usesSRV` seems to have similar logic to `getDNSType` and it is called earlier (in `validateMergedSDInputFields`), so it's a little confusing to have a check on the DNS type when it's only being explicitly set later. Wondering if that can be streamlined?

Reply:
Oh good point, it kind of looks like I almost wrote the same function twice